umbraco/Umbraco-CMS
# Nucache Performance fixes

### Prerequisites


If there's an existing issue for this PR then this fixes 

### Description
Update to 8.11.1
Fix binary bulk insert for SQLCE and ordering of properties.
Fix JsonSerializerSettings instantiation
Use ITransactableDictionary to decouple from bplustree
Move SQL from PublishedSnapshotService into DatabaseDataSource
Fix issue where Guid was not being mapped to Key in SQL Select for content types, it was being mapped to UniqueId.
Add component to automatically rebuild the cmsContentNu table when switching serializers.

Appears to work now @Shazwazza 

How to test?
Add to AppSettings in web.config for MessagePack
    `    <add key="Umbraco.Web.PublishedCache.NuCache.Serializer" value="MsgPack"/>`

Add to Appsettings in web.config for JSON serializer (defaults to JSON if no key)
`        <add key="Umbraco.Web.PublishedCache.NuCache.Serializer" value="JSON"/>`

If the Serializer has changed vs the state of the database , a rebuild of the published cache will occur.
Boot should be faster with MessagePack on subsequent boots when app_data/Temp/NuCache is empty.
To check what serializer the cmsContentNu table was serialized using:

`SELECT TOP (1) [key]
      ,[value]
      ,[updated]
  FROM [dbo].[umbracoKeyValue]
  where [key] = 'Umbraco.Web.PublishedCache.NuCache.Serializer'`
If no record = JSON.
